File: ios/TTF/TTF/Services/BackgroundCoverage.swift
import Foundation

/// Fire-and-forget coverage seeding — mirrors web `backgroundCoverage.ts`.
enum BackgroundCoverage {
    static let defaultSearchRadiusM = 8_000
    static let restaurantSeedRadiusM = 1_000

    private static let pollInterval: Duration = .seconds(4)
    private static let pollTimeout: Duration = .seconds(90)

    /// Queue a background seed and optionally refresh UI when the job finishes.
    static func run(
        api: APIClient,
        lat: Double,
        lng: Double,
        radiusM: Int,
        onComplete: (@MainActor () async -> Void)? = nil
    ) {
        Task { @MainActor in
            do {
                let response = try await api.ensureCoverage(lat: lat, lng: lng, radiusM: radiusM)
                if response.status == "queued", let jobID = response.jobID {
                    try await pollJob(jobID: jobID, api: api)
                }
                if let onComplete {
                    await onComplete()
                }
            } catch {
                // Background work — failures are non-blocking for the user.
            }
        }
    }

    private static func pollJob(jobID: UUID, api: APIClient) async throws {
        let deadline = ContinuousClock.now + pollTimeout
        while ContinuousClock.now < deadline {
            try await Task.sleep(for: pollInterval)
            let job: CoverageJobStatus
            do {
                job = try await api.getCoverageJob(id: jobID)
            } catch {
                continue
            }
            if job.status == "succeeded" || job.status == "failed" || job.status == "skipped" {
                return
            }
        }
    }
}

File: web/src/lib/backgroundCoverage.ts
import { api } from "../api/client";

const POLL_INTERVAL_MS = 4_000;
const POLL_TIMEOUT_MS = 90_000;

const delay = (ms: number) => new Promise<void>((resolve) => setTimeout(resolve, ms));

/** Poll a coverage seed job until it finishes or times out. */
export async function pollCoverageJob(
  jobId: string,
  token: string,
): Promise<"succeeded" | "failed" | "timeout"> {
  const deadline = Date.now() + POLL_TIMEOUT_MS;
  while (Date.now() < deadline) {
    await delay(POLL_INTERVAL_MS);
    try {
      const job = await api.getCoverageJob(jobId, token);
      if (job.status === "succeeded") return "succeeded";
      if (job.status === "failed") return "failed";
    } catch {
      continue;
    }
  }
  return "timeout";
}

/** Fire-and-forget area seed; optional callback when the job finishes or area is already covered. */
export function runBackgroundCoverage(
  lat: number,
  lng: number,
  radius_m: number,
  idToken: string,
  onComplete?: () => void,
): void {
  void (async () => {
    try {
      const res = await api.ensureCoverage({ lat, lng, radius_m }, idToken);
      if (res.status === "covered") {
        onComplete?.();
        return;
      }
      if (res.status === "queued" && res.job_id) {
        await pollCoverageJob(res.job_id, idToken);
        onComplete?.();
      }
    } catch {
      // Background work — failures are non-blocking for the user.
    }
  })();
}
